### PR TITLE
fix(Android): modal insets not applying

### PR DESCRIPTION
**Pull request by alduzy**
## Description

This PR fixes the insets not applying on modal screens.
I brought back the missing logic, that was removed [here](https://github.com/software-mansion/react-native-screens/commit/85a4196ec0194c7b647fe86c16841bb52e4eb665#diff-49e80f92048eb21d46beae35985ce78f79217693be04df175e5e679878d0f2c7L155) to `InsetsObserverProxy`

Note: Without this change the insets are applied correctly after navigating to a screen with `presentation: formSheet` at least once.

## Changes

- removed unused `unregister` fun from InsetsObserverProxy
- added `unregisterOnView` that resembles the previously removed logic

### Before
Screenshot 2024-10-02 at 13 07 51

### After
Screenshot 2024-10-02 at 13 54 15

## Test code and steps to reproduce

- use `Stack Presentation` in the Example app
- set `headerShown: false` for the modal screen
- open the modal screen

## Checklist

- [x] Ensured that CI passes

### CODE REVIEW

**Review by kkafar — COMMENTED** [review]
The changes look promising, however let's make sure we test this change on various `stackPresentation` & combinations. Also it might be important to test this in `bottom-tabs` environment.

Please confirm if you have tested these cases.

**Comment by alduzy** [comment]
@kkafar tested it with Example app's "Stack Presentation" and "Bottom tabs and native stack" + some test cases and can confirm it works as expected.

**Review by kkafar — COMMENTED** [review]
I'm still worried, that removing `InsetObserverProxy` from the decor view might interfere with `formSheet` presentation.

Imagine case where you have tab navigator, each screen is a native stack,  you present formSheet in one of them & open keyboard. I see it not working. I'll try to test it as well before merging.

**Comment by alduzy** [comment]
@kkafar I prepared a snack that can be used to test that:

<details>
<summary>snack</summary>

```
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { NavigationContainer } from '@react-navigation/native';
import {
  createNativeStackNavigator,
  NativeStackNavigationProp,
} from '@react-navigation/native-stack';
import React from 'react';
import { Button, TextInput } from 'react-native';

type StackParams = {
  Home: undefined;
  Sheet: undefined;
  Modal: undefined;
  Tab?: undefined;
};

const Stack = createNativeStackNavigator<StackParams>();
const Tab = createBottomTabNavigator();
const InnerStack = createNativeStackNavigator<StackParams>();

function HomeScreen({
  navigation,
}: {
  navigation: NativeStackNavigationProp<StackParams>;
}) {
  return (
    <>
      <Button title="Open sheet" onPress={() => navigation.navigate('Sheet')} />
      <Button title="Open modal" onPress={() => navigation.navigate('Modal')} />
      <Button
        title="Go to tab navigator"
        onPress={() => navigation.navigate('Tab')}
      />
    </>
  );
}

function InputScreen({
  navigation,
}: {
  navigation: NativeStackNavigationProp<StackParams>;
}) {
  return (
    <>
      <TextInput placeholder="Input something" />
      <Button title="Dismiss" onPress={() => navigation.popTo('Home')} />
    </>
  );
}

function InnerStackNavigator() {
  return (
    <InnerStack.Navigator>
      <InnerStack.Screen name="Home" component={HomeScreen} />
      <InnerStack.Screen
        name="Sheet"
        component={InputScreen}
        options={{ presentation: 'formSheet' }}
      />
      <InnerStack.Screen
        name="Modal"
        component={InputScreen}
        options={{ presentation: 'modal' }}
      />
    </InnerStack.Navigator>
  );
}

function TabNavigator() {
  return (
    <Tab.Navigator>
      <Tab.Screen name="Inner Stack" component={InnerStackNavigator} />
    </Tab.Navigator>
  );
}

export default function App() {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <InnerStack.Screen name="Home" component={HomeScreen} />
        <InnerStack.Screen
          name="Sheet"
          component={InputScreen}
          options={{ presentation: 'formSheet' }}
        />
        <InnerStack.Screen
          name="Modal"
          component={InputScreen}
          options={{ presentation: 'modal' }}
        />
        <Stack.Screen name="Tab" component={TabNavigator} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}
```
</details>

**Review by kkafar — APPROVED** [review]
Quickly tested this, did not saw any regression, however this is sensitive code, which I do not understand completely.

We need to do some deeper research on inset management in our apps.

File: android/src/main/java/com/swmansion/rnscreens/InsetsObserverProxy.kt [hunk 1]
```diff
@@ -57,10 +57,8 @@ object InsetsObserverProxy : OnApplyWindowInsetsListener {
         }
     }
 
-    fun unregister() {
-        eventSourceView.get()?.takeIf { hasBeenRegistered }?.let {
-            ViewCompat.setOnApplyWindowInsetsListener(it, null)
-        }
+    fun unregisterOnView(view: View) {
+        ViewCompat.setOnApplyWindowInsetsListener(view, null)
```
Comment (kkafar) [inline android/src/main/java/com/swmansion/rnscreens/InsetsObserverProxy.kt:61]: Note for future me:

This change broke formSheet behaviour with keyboard.
By calling it this way, we remove inset listener from `view` even if `InsetObserverProxy` has not been registered, that means if someone else registered on `view` we overwrite the listener to null.

**Pull request merged**